### PR TITLE
Improvements and fixes to wasm transaction planner and wasm view server

### PR DESCRIPTION
**Pull request by zpoken**
- added wasm planner which allows more flexible creation of TransactionPlan on TS side 
- added logic of storing and reading advice when scanning blocks
- added function for generating ephemeral address 
- planner reads data from indexedDB directly from rust

### CODE REVIEW

File: crates/wasm/Cargo.toml
Comment (grod220) [inline crates/wasm/Cargo.toml:42]: Unused?

File: crates/wasm/src/lib.rs
Comment (grod220) [inline crates/wasm/src/lib.rs:87]: How does `OsRng` work in the web context? Does it just work?

File: crates/wasm/src/lib.rs
Comment (grod220) [inline crates/wasm/src/lib.rs:81]: `JsValue` is quite painful trying to understand what comes back from this function. We also need to start adding docstrings above every imported function so it helps the frontend more. Something like:

```rust
/// Generates a kind of "burner" address in which the user can use to maintain their anonymity without sharing their main address.
/// Returns: `u128`
#[wasm_bindgen]
pub fn get_ephemeral_address(full_viewing_key: &str, index: u32) -> JsValue {
```

and this would work for parameters too:

```rust
/// Scans block for notes, swaps...
/// Arguments: 
///     compact_block: `v1alpha1::CompactBlock`
/// Returns: `ScanBlockResult`
#[wasm_bindgen]
    pub fn scan_block_without_updates(&mut self, compact_block: JsValue) -> JsValue {
```

File: crates/wasm/src/planner.rs
Comment (grod220) [inline crates/wasm/src/planner.rs:119]: What is this and what is it used for?

File: crates/wasm/src/planner.rs
Comment (grod220) [inline crates/wasm/src/planner.rs:418]: Delete?

File: crates/wasm/src/view_server.rs
Comment (grod220) [inline crates/wasm/src/view_server.rs:543]: I noticed we use a lot of unwrap()'s. Also these patterns:

```rust
.ok()?.await.ok()?;
.ok().unwrap()
```

In general, we need to move to the standard `Result` pattern and not use unwraps unless they are right as we coerce them to a `JsValue` and encode them for javascript.

Also, doing `.ok()` on everything has downsides:
- We lose error information, making debugging harder
- `?` is usually used on Result types. Using it on an Option is confusing.
- This will cause a panic if a None is received.
- We are treating Options as Results.

File: crates/wasm/src/wasm_planner.rs
Comment (grod220) [inline crates/wasm/src/wasm_planner.rs:95]: Can't these be?:

```rust
input_value: u128,
into_denom: &str,
swap_claim_fee: u128,
claim_address: &str,
```

Whenever possible, we should not do JsValue if it is allowed. You may have to test this.

File: crates/wasm/src/wasm_planner.rs
Comment (grod220) [inline crates/wasm/src/wasm_planner.rs:109]: What's the purpose of this function?

File: crates/wasm/src/wasm_planner.rs
Comment (grod220) [inline crates/wasm/src/wasm_planner.rs:149]: I think we should have a `IndexDb` struct that has all of these helper functions in one place in another file.

**Review by grod220 — COMMENTED** [review]
Please also run `rustfmt` and `clippy` on the wasm crate 🙏

File: crates/wasm/src/lib.rs
Comment (hdevalence) [reply crates/wasm/src/lib.rs:87]: Yeah, I believe the crate has compile-time features that hook the platform APIs.

**Comment by conorsch** [comment]
Discussed during sprint planning meeting today. We're currently planning to release in one business day, on Monday, #2984. Is this PR QOL, meaning it's OK to let slip for Monday? @grod220 will ping Zpoken folks to get a sense of urgency. It's technically still possible to dogpile on this and get it in for Monday if necessary.

File: crates/wasm/src/view_server.rs
Comment (Valentine1898) [reply crates/wasm/src/view_server.rs:543]: In most cases we will have to use a rather massive construction with `map_err`
` .map_err(|err| JsError::new(&err.to_string()))?`
On this [issue](https://github.com/rustwasm/wasm-bindgen/issues/1742) , there is no more ergonomic solution yet

File: crates/wasm/src/wasm_planner.rs
Comment (Valentine1898) [reply crates/wasm/src/wasm_planner.rs:109]: This is part of the functionality for flexible TransactionPlan creation
An example of use can be found [here](https://github.com/penumbra-zone/wallet/blob/dev/src/services/ViewProtocolService.ts#L279C4-L279C4)
With this function we add Swap actions to planner

**Comment by grod220** [comment]
Take a look at this example. At the top level, all `#[wasm_bindgen]` functions need to return  `Result<JsValue, Error>` if consuming functions that return results or just `JsValue` if not using any functions that that return results (should be very rare). 

However, every other function in this crate should be using the `Result` type (in my example the custom `WasmResult`). This example uses `thiserror` crate to handle the mapping of all possible errors in the code. This means, error messages are retained, you can use the succinct `?` without mapping errors, and the code is responsible for result handling like normal Rust code.

```rust

use std::str::FromStr;

use indexed_db_futures::IdbDatabase;
use indexed_db_futures::prelude::OpenDbRequest;
use serde_wasm_bindgen::Error;
use thiserror::Error;
use wasm_bindgen::JsValue;
use wasm_bindgen::prelude::wasm_bindgen;
use web_sys::DomException;

use penumbra_keys::keys::{SeedPhrase, SpendKey};
use penumbra_proto::DomainType;
use penumbra_proto::serializers::bech32str;

// --------------- //
// DomException doesn't play nicely with thiserror's automated tooling and needs to be wrapped

#[derive(Debug)]
pub struct DomError(DomException);

impl std::fmt::Display for DomError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "DOM Exception: {:?}", self.0)
    }
}

impl std::error::Error for DomError {}

impl From<DomException> for WasmError {
    fn from(dom_exception: DomException) -> Self {
        WasmError::Dom(DomError(dom_exception))
    }
}
// --------------- //

pub type WasmResult<T> = Result<T, WasmError>;


#[derive(Error, Debug)]
pub enum WasmError {
    #[error("{0}")]
    Anyhow(#[from] anyhow::Error),

    #[error("{0}")]
    Dom(#[from] DomError),
}

impl From<WasmError> for serde_wasm_bindgen::Error {
    fn from(wasm_err: WasmError) -> Self {
        Error::new(wasm_err.to_string())
    }
}

#[wasm_bindgen]
pub async fn some_wasm_fn() -> Result<JsValue, Error> {
    let result = get_spend_key().await?;
    Ok(JsValue::from_str(&result))
}

pub async fn get_spend_key() -> WasmResult<String> {
    let db_req: OpenDbRequest = IdbDatabase::open_u32("penumbra", 12)?;
    // do some stuff
    let spend_key = spend_key_new("xyz")?;
    Ok(spend_key)
}


pub fn spend_key_new(seed_phrase: &str) -> WasmResult<String> {
    let seed = SeedPhrase::from_str(seed_phrase)?;
    let spend_key = SpendKey::from_seed_phrase(seed, 0);

    let proto = spend_key.to_proto();
    Ok(bech32str::encode(
        &proto.inner,
        bech32str::spend_key::BECH32_PREFIX,
        bech32str::Bech32m,
    ))
}

```

**Comment by Valentine1898** [comment]
@grod220 
I've completed a major refactoring of wasm 
- added documentation for all wasm_bindgen functions 
- unwrap() and ok() constructs are no longer used 
- all indexedDB interactions have been moved to a separate file, and some of the code is being reused

**Comment by Valentine1898** [comment]
@zbuc 
In this PR, I have already gotten rid of using unwrap() in wasm crate 
I suggest leaving my variant as it uses succinct `?` and allows for proper error handling in JS

File: crates/wasm/src/tx.rs
Comment (grod220) [inline crates/wasm/src/tx.rs:100]: If deprecated, can we delete? This applies to all the functions labelled with `/// deprecated`

File: crates/wasm/src/tx.rs
Comment (grod220) [inline crates/wasm/src/tx.rs:92]: I'm not sure the `string_to_fvk` helper function makes sense if it's only one line. Think it can be done in-line.

File: crates/wasm/src/wasm_planner.rs
Comment (grod220) [inline crates/wasm/src/wasm_planner.rs:49]: Same feedback as the others. I'm not sure the point of having `inner` helper functions here given the body of the helper function is so small.

File: crates/wasm/src/wasm_planner.rs
Comment (grod220) [reply crates/wasm/src/wasm_planner.rs:165]: Same feedback as before 👆

File: crates/wasm/src/wasm_planner.rs
Comment (grod220) [inline crates/wasm/src/wasm_planner.rs:139]: If you are doing .expect() on these, why not just have `get_swap_by_commitment` return a Result instead of an Option?

File: crates/wasm/src/storage.rs
Comment (grod220) [inline crates/wasm/src/storage.rs:147]: With `thiserror`, you shouldn't need to be doing `.map_err(WasmError::from)` anymore. Just put the error in the new Error struct.

File: crates/wasm/src/storage.rs
Comment (grod220) [inline crates/wasm/src/storage.rs:146]: I would not convert this to an Option as every consumer is treating it as a Result.

Comment (grod220) [reply crates/wasm/src/storage.rs:146]: Same feedback for the rest of these as well

**Review by grod220 — COMMENTED** [review]
Substantially better and more succinct with Error handling now fixed 🎉 

Some additional comments. Let me know if you wanted to pair on this to make the remaining fixes or if you wanted to continue on your own.

**Comment by grod220** [comment]
Also, it looks like there are some merge conflicts with main. Please rebase/merge to fix 🙏

File: crates/wasm/Cargo.toml
Comment (Valentine1898) [reply crates/wasm/Cargo.toml]: Align and alphabetical order  :white_check_mark: 
As for the full version, I use the version that uses in other penumbra crates

**Comment by Valentine1898** [comment]
> Substantially better and more succinct with Error handling now fixed 🎉
> 
> Some additional comments. Let me know if you wanted to pair on this to make the remaining fixes or if you wanted to continue on your own.

I think I can make the fixes myself

File: crates/wasm/src/storage.rs
Comment (Valentine1898) [reply crates/wasm/src/storage.rs]: I removed it even earlier

File: crates/wasm/src/wasm_planner.rs
Comment (Valentine1898) [reply crates/wasm/src/wasm_planner.rs:49]: This is only necessary in order to use WasmResult<()> in the helper function

File: crates/wasm/src/wasm_planner.rs
Comment (Valentine1898) [reply crates/wasm/src/wasm_planner.rs:165]: The point is that on the TS side we will be using mostly [proto](https://buf.build/penumbra-zone/penumbra) types to interact with wasm. 
And in this case, having an object of the [Address](https://buf.build/penumbra-zone/penumbra/docs/main:penumbra.core.crypto.v1alpha1#penumbra.core.crypto.v1alpha1.Address) type on the TS side, we will need to 
1. convert Address -> string on the TS side 
2. on the wasm side get an Address from string object
This leads to additional writing of specific conversion code, while the use of JsValue will be unified everywhere else

File: crates/wasm/src/storage.rs
Comment (Valentine1898) [reply crates/wasm/src/storage.rs:146]: I proceeded from the fact that the situation when we successfully read data from indexedDb by a given key, but there was no record by this key in indexedDB, should not lead to an error.

File: crates/wasm/src/tx.rs
Comment (Valentine1898) [inline crates/wasm/src/tx.rs:129]: For example, Option is used here when indexedDB reading

File: crates/wasm/Cargo.toml
Comment (Valentine1898) [reply crates/wasm/Cargo.toml]: ✅

File: crates/wasm/publish/publish-wasm.ts
Comment (Valentine1898) [reply crates/wasm/publish/publish-wasm.ts:1]: ✅

File: crates/wasm/src/error.rs
Comment (Valentine1898) [reply crates/wasm/src/error.rs]: ✅

File: crates/wasm/src/error.rs
Comment (Valentine1898) [reply crates/wasm/src/error.rs:12]: ✅

File: crates/wasm/src/keys.rs
Comment (Valentine1898) [reply crates/wasm/src/keys.rs:25]: I agree, I'll remove the `js_name`

Comment (Valentine1898) [reply crates/wasm/src/keys.rs]: ✅

File: crates/wasm/src/keys.rs
Comment (Valentine1898) [reply crates/wasm/src/keys.rs:35]: Yes
image

File: crates/wasm/src/keys.rs
Comment (Valentine1898) [reply crates/wasm/src/keys.rs]: Removed `pub` wherever it was unnecessary

File: crates/wasm/src/planner.rs
Comment (Valentine1898) [reply crates/wasm/src/planner.rs:464]: Actually this code is copied from [here](https://github.com/penumbra-zone/penumbra/blob/main/crates/view/src/planner.rs), we just can't reuse the code from view crate because it is not compatible with wasm

File: crates/wasm/src/tx.rs
Comment (Valentine1898) [reply crates/wasm/src/tx.rs:97]: That means it still uses [penumbra-zone/wallet](https://github.com/penumbra-zone/wallet) , so we can't remove it yet.

**Comment by Valentine1898** [comment]
I think I've fixed all the comments except those related to these two issues 
1. Should functions reading from indexedDb return `WasmResult<Option<SwapRecord>> `
2. most helper functions are meaningless, they are only needed so that `#[wasm_bindgen]` functions return `Result<JsValue, Error>` and we can use succinct `?`

File: crates/wasm/src/keys.rs
Comment (grod220) [reply crates/wasm/src/keys.rs:35]: Amazing 🎉

File: crates/wasm/src/storage.rs
Comment (grod220) [reply crates/wasm/src/storage.rs:146]: It feels easier to make this a Result as we are nearly always expecting data to be there. It looks like there is only one Option case that could easily work the same as a Result. If you feel strongly otherwise, I can live with it though.

File: crates/wasm/src/wasm_planner.rs
Comment (grod220) [reply crates/wasm/src/wasm_planner.rs:49]: In that case we should do this:

```rust
// In wasm_planner.rs
pub fn memo(&mut self, memo: JsValue) -> WasmResult<()> {
    let memo_proto: MemoPlaintext = serde_wasm_bindgen::from_value(memo)?;
    let _ = self.planner.memo(memo_proto.try_into()?);
    Ok(())
}

// in error.rs
impl From<WasmError> for JsValue {
    fn from(error: WasmError) -> Self {
        JsError::from(error).into()
    }
}
```

And then we can use WasmResult everywhere, even the `#[wasm_bindgen]` top level functions. It will remove the need of using the `inner` pattern too. Helper functions are not bad, but we should use them when necessary and not to please the `wasm_bindgen` gods.

File: crates/wasm/Cargo.toml
Comment (grod220) [reply crates/wasm/Cargo.toml:32]: More information on this: https://users.rust-lang.org/t/psa-please-specify-precise-dependency-versions-in-cargo-toml/71277

File: crates/wasm/src/storage.rs
Comment (Valentine1898) [reply crates/wasm/src/storage.rs:146]: Another point is that when initially reading from indexedDB, we get `Option<JsValue> `
```rust
 let read_result: Option<JsValue> = store
            .get_owned("chain_parameters")?
            .await?;
``` 
This means we'd need to convert `Option -> Result`,  and it doesn't seem to be very convenient
For some functions like `get_chain_parameters()` and `get_fmd_parameters()`, we do expect that either the record exists or something is wrong. 
But in other cases, there is a high probability that we will successfully read from indexedDb, but the data will not be there. And in the future, if we want more interaction with indexedDb, there will be more logic similar to 
```rust
  let option = storage.get_note_by_nullifier().await?;
    match option {
        None => {
            // forget nullifier
        }
        Some(note_record) => {
            // Mark record_note as spent
        }
    }
```

File: crates/wasm/Cargo.toml
Comment (Valentine1898) [reply crates/wasm/Cargo.toml]: I agree, fixed it

File: crates/wasm/src/tx.rs
Comment (Valentine1898) [reply crates/wasm/src/tx.rs:100]: ✅

File: crates/wasm/src/wasm_planner.rs
Comment (Valentine1898) [reply crates/wasm/src/wasm_planner.rs]: I removed some of the helper functions that seemed redundant. ✅

File: crates/wasm/src/keys.rs
Comment (grod220) [inline crates/wasm/src/keys.rs:1]: Think this file has some helper functions that are small enough to be moved into the main function. It's similar feedback as the other. Will have to move `WasmResult` to the the top level function.

**Review by grod220 — APPROVED** [review]
Wow! 🎊 

Though this PR took a lot of iterations, think the wasm crate is in a _much_ better place with these new patterns. Going to approve, but think there's one file where the helpers are small enough to move into the main function. After you do that, feel free to merge 👍

File: crates/wasm/src/keys.rs
Comment (Valentine1898) [reply crates/wasm/src/keys.rs:1]: ✅

**Comment by Valentine1898** [comment]
@conorsch can merge this PR?

**Comment by conorsch** [comment]
There's just one more conflict, on the cargo lock file, that's blocking merge. I'll fix that now, and get it in!

**Comment by conorsch** [comment]
Turns out rebasing this PR on latest main was not straightforward. After an hour of unsuccessful interactive rebasing, I ended up performing a mixed reset on top of main, selectively adding only the wasm-related changes, and discarding the rest. That approach looks to have succeeded in preserving the new content of the PR while catching up the base with latest main.

In case I made a mistake, I made sure to push a "backup" branch to the penumbra remote: https://github.com/penumbra-zone/penumbra/tree/wasm_transaction_planner-backup That branch captures the state of this PR prior to my rebasing.

There's only one commit on this PR now, including the entirety of changes over the course of review.  Tests and linters pass locally, so I think we're in good shape.

@grod220 and @Valentine1898: given your familiarity with the changes here, can you take one more look at the diff and confirm it conforms to expectations? If so, we're good to merge!

**Review by conorsch — APPROVED** [review]

**Comment by Valentine1898** [comment]
Oh, I didn't think merge would be a problem. 
I reviewed the code again, it's all right.

**Review by grod220 — APPROVED** [review]
@Valentine1898, please merge 🙏

**Comment by Valentine1898** [comment]
I can't merge, I don't have permission to do that

**Comment by grod220** [comment]
@Valentine1898 oh! Didn't realize. I feel like an approved PR by penumbra team member should enable external folks to merge 🤔 . I'll do it, all good.

**Pull request merged**
